core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -724,6 +777,62 @@ private[spark] class ExecutorAllocationManager(
     }
 
     /**
+     * Calculate the maximum no. of concurrent tasks that can run currently.
+     */
+    def getMaxConTasks(): Int = {
+      val stagesByJobGroup = stageIdToNumTasks.groupBy(x => jobIdToJobGroup(stageIdToJobId(x._1)))
```

Comment:
I think this needs a comment explaining why you need to look at stages at all -- its not obvious why its necessary.  (at first I was going to suggest the number of tasks left in a stage shouldn't matter, but then realized that it could in some scenarios)